File: functions/exchange_rate/application/integrity_service.py
from datetime import datetime, timedelta

from firebase_functions import logger

from exchange_rate.adapter.output.database.models.exchange_rate import ExchangeRateAR


class IntegrityService:
    _NUMBER_OF_ROWS = 24 * 5

    def check_integrity(self):
        logger.info("Checking database integrity...")

        failed_days = []

        start_date = datetime(year=2025, month=2, day=20, hour=0, minute=0, second=0)
        current_date = start_date
        now = datetime.now().replace(minute=0, second=0, microsecond=0)

        while current_date < now:
            rows_count = ExchangeRateAR.select().where((ExchangeRateAR.datetime >= current_date) & (ExchangeRateAR.datetime < (current_date + timedelta(hours=24)))).count()
            if rows_count != self._NUMBER_OF_ROWS:
                failed_days.append(current_date)
                logger.error(f"Integrity check failed for {current_date}: expected {self._NUMBER_OF_ROWS} rows, found {rows_count}")

            current_date += timedelta(days=1)

        if failed_days:
            logger.info(f"Integrity check completed. Failed days: {failed_days}")

            for failed_day in failed_days:
                rows = ExchangeRateAR.select().where((ExchangeRateAR.datetime >= failed_day) & (ExchangeRateAR.datetime < (failed_day + timedelta(hours=24))))
                missing_hours = []
                for hour in range(0, 24):
                    if rows.where(ExchangeRateAR.datetime == failed_day.replace(hour=hour)).count() != 5:
                        missing_hours.append(hour)

                logger.error(f"Missing hours for {failed_day}: {missing_hours}")
        else:
            logger.info("Integrity check completed. No issues found.")

File: functions-ts/src/exchange-rate/application/integrity-service.ts
import * as logger from 'firebase-functions/logger'
import { addDays, addHours } from 'date-fns'
import { SupabaseClient } from '@supabase/supabase-js'
import { Database } from '../../../generated/database.types'
import { assertPresent } from 'ts-extras'

export class IntegrityService {
  private NUMBER_OF_ROWS = 24 * 5

  constructor(private readonly db: SupabaseClient<Database>) {}

  async checkIntegrity() {
    logger.info('Checking integrity service...')

    const failedDays: Date[] = []

    const startDate = new Date(2025, 2, 20, 0, 0, 0, 0)
    let currentDate = startDate
    const now = new Date()
    now.setMilliseconds(0)
    now.setSeconds(0)
    now.setMinutes(0)
    now.setHours(0)

    while (currentDate < now) {
      const { count: rowsCount } = await this.db
        .schema('private')
        .from('exchange_rates')
        .select('registered_at', { count: 'exact', head: true })
        .gte('registered_at', currentDate.toISOString())
        .lt('registered_at', addHours(currentDate, 24).toISOString())

      assertPresent(rowsCount)

      if (rowsCount !== this.NUMBER_OF_ROWS) {
        failedDays.push(currentDate)
        logger.warn(
          `Integrity check failed for ${currentDate}: expected ${this.NUMBER_OF_ROWS} rows, found ${rowsCount}`,
        )
      }

      currentDate = addDays(currentDate, 1)
    }

    if (failedDays.length > 0) {
      logger.warn(
        `Integrity check completed. Failed days: ${failedDays.join(',')}`,
      )

      for (const failedDay of failedDays) {
        const rows = this.db
          .schema('private')
          .from('exchange_rates')
          .select('registered_at', { count: 'exact', head: true })
          .gte('registered_at', failedDay.toISOString())
          .lt('registered_at', addHours(failedDay, 24).toISOString())

        const missingHours = []

        for (const hour of Array.from({ length: 24 }, (v, t) => t)) {
          failedDay.setHours(hour)

          if (
            (await rows.eq('registered_at', failedDay.toISOString())).count !==
            5
          ) {
            missingHours.push(hour)
          }
        }

        logger.error(
          `Missing hours for ${failedDay}: [${missingHours.join(',')}]`,
        )
      }
    } else {
      logger.info('Integrity check completed. No issues found.')
    }
  }
}
